stake-pool/program/src/processor.rs

```diff
@@ -2162,16 +2220,88 @@ impl Processor {
         Ok(())
     }
 
+    /// Processes [SetWithdrawalFee](enum.Instruction.html).
+    fn process_set_withdrawal_fee(
+        program_id: &Pubkey,
+        accounts: &[AccountInfo],
+        fee: Fee,
+    ) -> ProgramResult {
+        let account_info_iter = &mut accounts.iter();
+        let stake_pool_info = next_account_info(account_info_iter)?;
+        let manager_info = next_account_info(account_info_iter)?;
+        let clock_info = next_account_info(account_info_iter)?;
+        let clock = &Clock::from_account_info(clock_info)?;
+
+        check_account_owner(stake_pool_info, program_id)?;
+        let mut stake_pool = try_from_slice_unchecked::<StakePool>(&stake_pool_info.data.borrow())?;
+        if !stake_pool.is_valid() {
+            return Err(StakePoolError::InvalidState.into());
+        }
+
+        stake_pool.check_manager(manager_info)?;
+
+        if stake_pool.last_update_epoch < clock.epoch {
+            return Err(StakePoolError::StakeListAndPoolOutOfDate.into());
+        }
```

Comment:
Actually, we kind of want this. Just like for `set_fee`, if the epoch has changed, we want the manager to only be able to `set_withdrawal_fee` for the following epoch, not the current one.

Reply:
right, agreed with @jon-chuang. I didn't realize before but without this check it's possible for a manager to set the withdrawal fee for the current epoch instead of the next one by calling `set_withdrawal_fee` before anyone has called `update_stake_pool_balance`

Reply:
we're working on a new branch with the `Fee` struct reworked so that we don't need checked arithmetic operations. In brief here's the proposal:

- `numerator` and `denominator` are private `u64` fields
- following restrictions checked on `Fee` initialization: `denominator` >= 1, `numerator` <= `denominator`, `denominator` < some arbitrary precision such that multiplication and division will never overflow `u64`. We're thinking `1_000_000_000`
- impl `PartialOrd` and overload `std::ops::Mul`

so, basically changing `Fee` into a constrained fraction

Reply:
Ah right, sounds good then!